cmd/generic-handlers.go

```diff
@@ -469,3 +477,90 @@ func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 	h.handler.ServeHTTP(w, r)
 }
+
+// logginHandler logs all HTTP requests and responses
+// and write the log to a io.Writer object.
+type loggingHandler struct {
+	writer    io.Writer
+	writeLock *sync.Mutex
+	handler   http.Handler
+}
+
+func (h loggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	const timeFormat = "2006-01-02 15:04:05 -0700"
+	const recordLen = 32 * 1024
+
+	// Generate short random request ID
+	reqID := fmt.Sprintf("%X", time.Now().UnixNano()%1e8)
+
+	reqTime := time.Now().Format(timeFormat)
+
+	// Setup a http request body recorder
+	reqBodyRecorder := &recordReader{Reader: r.Body, recLen: recordLen}
+	r.Body = ioutil.NopCloser(reqBodyRecorder)
+
+	// Setup a http response body recorder
+	rec := httptest.NewRecorder()
+	respBodyRecorder := &recordResponseWriter{ResponseWriter: rec, recLen: recordLen}
+
+	h.handler.ServeHTTP(respBodyRecorder, r)
+	respTime := time.Now().Format(timeFormat)
+
+	// Copy the captured response headers to the new response
+	for k, v := range rec.Header() {
+		w.Header()[k] = v
+	}
+	w.WriteHeader(rec.Code)
+	// Copy the captured response body to the ew response
+	_, wErr := rec.Body.WriteTo(w)
+	if wErr != nil {
+		errorIf(wErr, "Cannot dump HTTP request")
+		return
+	}
+
+	// Log HTTP Request
+	requestDump, dumpReqErr := httputil.DumpRequest(r, false)
+	if dumpReqErr != nil {
+		errorIf(dumpReqErr, "Cannot dump HTTP request")
+		return
+	}
+	requestDump = append(requestDump, reqBodyRecorder.Data()...)
+
+	// Log HTTP Response
+	respDump, dumpRespErr := httputil.DumpResponse(rec.Result(), false)
+	if dumpRespErr != nil {
+		errorIf(dumpRespErr, "Cannot dump HTTP response")
+		return
+	}
+	respDump = append(respDump, respBodyRecorder.Data()...)
+
+	logText := fmt.Sprintf("[%s] [%s] >> %s\n\n[%s] [%s] << %s\n\n",
+		reqID, reqTime, string(requestDump),
+		reqID, respTime, string(respDump),
+	)
+
+	h.writeLock.Lock()
+	h.writer.Write([]byte(logText))
```

Comment:
As we log request/response together, till the response is returned, the request is not logged.  Similarly if response is not returned, the request is not at all logged.